HPCC-24303 Improve capture of ESP stats

kenrowland opened the pull request:
Added support in ecm service definition to automatically generate execution profile histograms

Signed-off-by: Ken Rowland Kenneth.Rowland@lexisnexisrisk.com

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Cloud-compatibility
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

Commit:
HPCC-24303 Improve capture of ESP stats

Added support in ecm service definition to automatically generate
execution profile histograms

Signed-off-by: Ken Rowland Kenneth.Rowland@lexisnexisrisk.com

afishbeck commented on tools/hidl/hidlcomp.cpp:
```diff
@@ -7034,38 +7114,87 @@ void HIDLcompiler::Process()
+
+        //
+        // At the HIDL top level, execution profiling is enable if any service is enabled
+        executionProfilingEnabled |= serviceProfileExecutionEnabled;
```
Interesting way of updating the higher level flag.  I wonder, is this faster or clearer than the more explicit:
```
if (!executionProfilingEnabled && serviceProfileExecutionEnabled)
    executionProfilingEnabled = true;
```
?

kenrowland replied:
Since this is build time, performance is not as critical, but I think the |= is faster. If you'd like the more explicit version, I can change it.

afishbeck commented on tools/hidl/hidlcomp.cpp:
```diff
@@ -7034,38 +7114,87 @@ void HIDLcompiler::Process()
+
+    if (executionProfilingEnabled)
+        outs("#include \"espcommon.hpp\"\n");
+        outs("#include \"jmetrics.hpp\"\n");
```
Did you mean?:
```
   if (executionProfilingEnabled)
   {
        outs("#include \"espcommon.hpp\"\n");
        outs("#include \"jmetrics.hpp\"\n");
   }
```

kenrowland replied:
Yes, probably. Good catch, although not harmful. Better to fix for sure.

afishbeck commented on esp/espcommon/espcommon.cpp:
```diff
@@ -0,0 +1,71 @@
+// function to return a set of buckets for an execution profile definitions
+ESPCOMMON_API std::vector<__uint64> getExecutionProfileBuckets(const std::string &profileOptions)
+{
+    std::vector<std::string> executionProfileDef = splitString(profileOptions, ',');
```
Trivial but FYI we do have existing routines that do this, it would just mean using a StringArray for executionProfileDef instead of a std::vector.  And it would mean not needing to create the splitString function.

```
StringArray executionProfileDef ;
executionProfileDef.appendList(profileOptions.c_str(), ",");
```

kenrowland replied:
If it's OK, I'd like to stick with the std lib types.

afishbeck reviewed (commented):
@kenrowland some fairly minor comments.

kenrowland commented on esp/espcommon/espcommon.cpp:
```diff
@@ -0,0 +1,71 @@
+}
+
+// Converts the bucket limit in logical units to the standard units used by the metric framework
+__uint64 convertToBucketLimit(const std::string &units, const std::string &value)
```
I don't see it being part of jmetrics at this time. If other components accept a string to define the histogram buckets, we can revisit moving it.

kenrowland commented on tools/hidl/hidlcomp.cpp:
```diff
@@ -384,12 +386,30 @@ void indentOutf1(int inc, const char* fmt,...) __attribute__((format(printf,2,3)
 
+bool validateProfileExecutionOptions(std::string &options)
+{
+    auto pos = options.find_first_of(',');
```
Changed type to size_t

kenrowland commented:
@richardkchapman @ghalliday Please merge

afishbeck commented:
@kenrowland I would say that the title of your PR could be a lot more descriptive.  "Improve capture of ESP stats".  Is pretty generic and not actually what you are doing.  What you are really doing is adding support for the instrumentation of some metrics to ESDL / HIDL.

ghalliday commented on tools/hidl/hidlcomp.h:
```diff
@@ -989,6 +996,32 @@ class EspMethodInfo
+        return true;
+    }
+
+    std::string getExecutionProflingMetricName()
```
typo: profiling

ghalliday commented on esp/espcommon/espcommon.cpp:
```diff
@@ -0,0 +1,71 @@
+}
+
+
+//// function to return the scale factor based on input units
```
Clean up unused code.

ghalliday reviewed (commented):
@kenrowland broadly lookgs good.  A few areas that need looking at
- it needs more information in the jira about exactly what has been implemented
- this should have a documentation impact.  A new executionProfilingOptions has been option to the service definitions, there needs to be a description of the syntax and the effect.
- Some concern that use of std::string is duplicating functions that already exist, or make function prototypes ineffient
- Some scope for reducing the size of the generated code.
- Cleaning up some out of date comments and warnings

kenrowland commented:
@ghalliday Please review changes. I opened a sub-task Jira to which I'd like to move this. Once the review is done, I'd like to squash these changes and rebase them to another branch matching the new Jira (which is more descriptive). Then close this PR.

ghalliday commented on tools/hidl/hidlcomp.cpp:
```diff
@@ -5760,7 +5751,7 @@ void EspServInfo::write_esp_binding()
             outf("\tprofileOptions = std::string(\"%s\");\n", mthi->getExecutionProfilingOptions().c_str());
             outs("\tbuckets = getExecutionProfileBuckets(profileOptions);\n");
-            outf("\t%s = hpccMetrics::registerCyclesToNsScaledHistogramMetric(\"%s\", \"\", buckets);\n", mthi->getExecutionProflingMetricName().c_str(), mthi->getName());
+            outf("\t%s = hpccMetrics::registerCyclesToNsScaledHistogramMetric(\"%s\", \"\", buckets);\n", mthi->getExecutionProfilingMetricName().c_str(), mthi->getName());
```
Missed the comment above about combining the functions - possibly because it wasn't attached to a line that has changed.

kenrowland replied:
Expanded the comment above the loop to indicate that its initializing the histogram metric for each method with execution profiling enabled.

ghalliday commented on tools/hidl/hidlcomp.cpp:
```diff
@@ -7034,38 +7122,89 @@ void HIDLcompiler::Process()
+    // metrics execution profiling requires the memory header. Search the services and/or methods to determine
+    // if enabled in any. If so, write the include statement
+    EspServInfo *si;
+    bool profilingEnabled = false;
```
Missed this comment

kenrowland replied:
Added -Wunused-variable to the compile options. Resulted in a number of other unused variable warnings that I also removed.

ghalliday reviewed (commented):
@kenrowland looking good.  Two missed comments from before (commoning up generated code and unused variables)

kenrowland commented:
@ghalliday Once complete, to make the PR title more descriptive, I can either change the title when rebasing, or close this PR and make a new one under a different Jira (https://track.hpccsystems.com/browse/HPCC-28576). Or, we can just leave it alone and not worry about it.

ghalliday commented:
@kenrowland I think you misunderstood my last comments.  There were 2 comments from previous PRs you misssed - it wasn't a comment about comments in the code.
You should change the commit message when squashing this PR - it should be something that makes sense on its own in a change log.  There is no need to open a new PR.

kenrowland commented on tools/hidl/hidlcomp.cpp:
```diff
@@ -5702,11 +5742,34 @@ void EspServInfo::write_esp_binding()
+    {
+        if (mthi->isExecutionProfilingEnabled())
+        {
+            outf("\tprofileOptions = std::string(\"%s\");\n", mthi->getExecutionProfilingOptions().c_str());
```
Converted to a function in espcommon. That is one of the intended goals, so this is a good first step. I will test the generated code to make sure it works in a functioning ESP.

kenrowland commented on tools/hidl/hidlcomp.cpp:
```diff
@@ -7034,38 +7122,89 @@ void HIDLcompiler::Process()
+    // if enabled in any. If so, write the include statement
+    EspServInfo *si;
+    bool profilingEnabled = false;
+    StrBuffer tmp;
```
Removed

ghalliday replied:
tmp is also unused and hasn't been reported.  (Because it has a non-trivial constructor.)

kenrowland commented on tools/hidl/hidlcomp.cpp:
```diff
@@ -7034,38 +7122,89 @@ void HIDLcompiler::Process()
+                    mthi->setEnableExecutionProfiling(true);
+                    if (!mthi->setExecutionProfilingOptions(methodProfileExecutionEnabled ? methodProfilingOptions.str() : serviceProfilingOptions.str()))
+                    {
+                        throw "Invalid execution profile options detected";
```
I followed the standard used elsewhere in the code. If you want, I can build a string with the name of the method and throw that. You are right that it may leak, but it's probably not an issue since this is a tool that only runs at build time. Please let me know if you want that change before final approval.

kenrowland reviewed (commented):
@ghalliday  I did misunderstand your last comments. The comments to which you referred were collapsed. I addressed them in the latest commit. Please see my latest comments. I'm not able to request a re-reviews, so please take a look.

ghalliday commented on esp/espcommon/espcommon.cpp:
```diff
@@ -37,9 +37,8 @@ __uint64 convertToBucketLimit(const char *units, const char *value)
-
 // function to return a set of buckets for an execution profile definitions
-ESPCOMMON_API std::vector<__uint64> getExecutionProfileBuckets(const std::string &profileOptions)
+std::vector<__uint64> getExecutionProfileBuckets(const std::string &profileOptions)
```
minor: would avoid creating a std::string if this was a const char * parameter

ghalliday replied:
could also make it static if not exported.

ghalliday commented on tools/hidl/CMakeLists.txt:
```diff
@@ -33,23 +33,23 @@ add_custom_command ( OUTPUT ${CMAKE_CURRENT_BINARY_DIR}/hidllex.cpp
 set_source_files_properties (${CMAKE_CURRENT_BINARY_DIR}/hidlgram.cpp PROPERTIES COMPILE_FLAGS -fno-strict-aliasing)
 
-ADD_DEFINITIONS( -D_CONSOLE)
+ADD_DEFINITIONS( -D_CONSOLE -Wunused-variable)
```
The following are the preferred options:

```
  -Wall -Wextra -Wno-switch -Wno-unused-parameter -Werror -Wno-error=delete-non-virtual-dtor
```
so that warnings are reported as errors.  (e.g see CMakeList.txt for jlib)

kenrowland replied:
I added the options as indicated. There were a number of additional errors that had to be corrected. Please see hidllex.cpp. In that file a method and it's prototype were commented out since they weren't used. However, there are a number of instances in the code that use the definition that controlled its inclusion. Not knowing this part of the code, I decided to comment it out for now.

ghalliday reviewed (commented):
A few (hopefully) final comments

kenrowland commented:
@AttilaVamos I am looking into this. It has to do with yylex generated code.

ghalliday commented on tools/hidl/hidlgram.y:
```diff
@@ -1676,29 +1680,34 @@ void check_param(void)
-        }
+            break;
+        default:
+            invalid_return_type();
```
trivial: Should have a break after this statement to stop coverity complaining.

ghalliday reviewed (approved):
Looks good.  One trivial comment.  Please sqaush.  Re-requesting tony to look at the changes since quite a lot has happened since he last reviewed.

afishbeck commented on esp/espcommon/espcommon.cpp:
```diff
@@ -0,0 +1,58 @@
+}
+
+// Registers a profiling metric for an ESP service method
+ESPCOMMON_API std::shared_ptr<hpccMetrics::ScaledHistogramMetric> registerProfilingMetric(const char *histogramMetricName, const char *desc, const char *profilingOption)
```
A bit inconsistent with singular vs plural name for profilingOption(s).

afishbeck commented on tools/hidl/hidlcomp.cpp:
```diff
@@ -384,12 +387,22 @@ void indentOutf1(int inc, const char* fmt,...) __attribute__((format(printf,2,3)
     voutf(fmt,args);
 }
 
+void validateProfileExecutionOptions(std::string &options)
```
For a string validation function, a comment showing the expected format would be helpful.

afishbeck commented on tools/hidl/hidlcomp.h:
```diff
@@ -989,6 +996,29 @@ class EspMethodInfo
+
+    std::string getExecutionProfilingMetricName()
+    {
+        std::string name("p");
```
You could save the qualified name in the EspMethodInfo class rather than building it every time.

afishbeck reviewed (commented):
A few fairly trivial comments.

kenrowland commented:
@afishbeck @ghalliday Comments addressed. There are three additional changes in the latest commit
- Cleaned up header file usage in espcommon
- Added statement to hidl lex file to prevent generation of yyunput to fix a build problem in CentOS (caused by new build flags)
- Added validation of execution profiling options
I have re-requested a review to make sure these are covered.

afishbeck commented on tools/hidl/hidlcomp.h:
```diff
@@ -989,6 +997,34 @@ class EspMethodInfo
+    {
+        enableExecutionProfiling = enable;
+
+        if (enable)
```
You only call this once, but still prefer to code defensively.  Should check to see if executionProfilingHistogramVariableName is empty befroe appending, or if there were a reason to update then clear it.

kenrowland replied:
Changed the method to only turn on profiling (once). There is no reason to support allowing it to be turned off.

afishbeck commented on tools/hidl/hidlcomp.h:
```diff
@@ -989,6 +997,34 @@ class EspMethodInfo
+    const std::string &getExecutionProfilingOptions(){return executionProfilingOptions;}
+    void setExecutionProfilingOptions(const char *options)
+    {
+        executionProfilingOptions = std::string(options);
```
Curious why you can't simply?:

executionProfilingOptions = options;

kenrowland replied:
I generally am more explicit in areas like this so that when you are looking at the code you'd know that a std::string was being constructed. I changed it to be more concise.

afishbeck reviewed (commented):
Some comments.

kenrowland commented on tools/hidl/hidlcomp.h:
```diff
@@ -989,6 +997,34 @@ class EspMethodInfo
+        executionProfilingOptions = std::string(options);
+        auto noSpaceEnd = std::remove(executionProfilingOptions.begin(), executionProfilingOptions.end(), ' ');
+        executionProfilingOptions.erase(noSpaceEnd, executionProfilingOptions.end());
+        validateProfileExecutionOptions(executionProfilingOptions);
```
You are correct, I do remove the spaces before calling the validation. However, the function could be called by someone else. I removed the check and made a note that the caller must remove any spaces first.

afishbeck commented on tools/hidl/hidlcomp.cpp:
```diff
@@ -399,11 +399,7 @@ void validateProfileExecutionOptions(std::string &options)
-
-    if (options.find_first_of(' ') != std::string::npos)
-        throw "Execution profiling option may not include blanks";
+    // Note, caller must remove any spaces
```
This is correct, but this wasn't exactly the intention of my comment.  Caller must remove spaces in order for there to be consistency, not just to pass this test.

Your syntax accepts spaces, it's an internal detail that they are stripped out before use.  If another place calls this validate and it doesn't work because spaces aren't removed then it will be confusing for users.

afishbeck reviewed (approved):
Approved.  @kenrowland trivial comment explaining the intent of my prior comment.

ghalliday commented:
@kenrowland this is failing the macos12 build for the following reasons:

```
/Users/runner/work/HPCC-Platform/HPCC-Platform/build/tools/hidl/hidllex.cpp:2117:16: error: function 'yyinput' is not needed and will not be emitted [-Werror,-Wunneeded-internal-declaration]
    static int yyinput (void)
               ^
/Users/runner/work/HPCC-Platform/HPCC-Platform/tools/hidl/hidlcomp.cpp:5537:15: error: format string is not a string literal [-Werror,-Wformat-nonliteral]
        voutf(fmt, args);
              ^~~
```

ghalliday commented on tools/hidl/hidlcomp.cpp:
```diff
@@ -4428,13 +4428,13 @@ void EspMessageInfo::write_esp()
             outf("\tdouble clientVer = ctx ? ctx->getClientVersion() : -1;\n");
         //attributes first
-        int attrCount=0;
+        int attrribCount=0;
```
Not sure how renaming these variables is related to fixing the warnings, and this seems to accidentally have a double rr in it (which makes it heard to read).

ghalliday commented on tools/hidl/hidlcomp.cpp:
```diff
@@ -395,12 +395,42 @@ void indentOutf1(int inc, const char* fmt, ...)
+    int value, maxValue = 0;
+    for (unsigned i=1; i<optionValues.size(); ++i)
+    {
+        value = std::stoi(optionValues[i]);
```
picky: easier to read if value is also declared here, rather than wondering why value might be used after the loop, and is it a problem it isn't initialised.

ghalliday reviewed (commented):
A couple of minor comments.  Only one really needs fixing. (attrrib)

ghalliday commented:
@kenrowland looks like you only addressed one of the warnings/errors from macos.  The other one is:

```
/Users/runner/work/HPCC-Platform/HPCC-Platform/tools/hidl/hidlcomp.cpp:5537:15: error: format string is not a string literal [-Werror,-Wformat-nonliteral]
        voutf(fmt, args);
              ^~~
```

It requires adding __attribute__((format(printf, ....)) to various functions.

kenrowland commented:
@ghalliday @richardkchapman Please merge.

ghalliday commented:
@kenrowland what documentation impact does this have?  No checkboxes were ticked, but there must be some.  @JamesDeFabia  where is the syntax for esdl files defined?

kenrowland commented:
@ghalliday I don't think we are ready to add this to the documentation yet. It would be part of a comprehensive set of documentation on metrics. My plan was to have a separate Jira at a later date.

ghalliday commented:
It is worth creating a jira for the documentation, and including the information from each stage as it is implemented, to make sure that information isn't lost.
